---
title: "coralMicrobiomes"
author: "Matthew J. Neave"
date: "April 22, 2015"
output: pdf_document
---


This is an R Markdown document detailing the statistical and graphical steps for reproducting the results in:

Neave, M.J., Rachmawati, R., Xun, L., Michell, C.T., Barber, P.H., Bourne, D.G., McCulloch, M.T., Apprill, A., Voolstra, C.R. A global microbiome analysis reveals that closely related corals exhibit fine-scale differences in their association with Endozoicomonas symbionts. 

### Load required libraries  

```{r setup, message=FALSE}
library("phyloseq"); packageVersion("phyloseq")
library("ggplot2"); packageVersion("ggplot2")
library("plyr"); packageVersion("plyr")
library("vegan"); packageVersion("vegan")
library("grid"); packageVersion("grid")
library("knitr"); packageVersion("knitr")
library("clustsig"); packageVersion("clustsig")
library('ape'); packageVersion("ape")
library('RColorBrewer'); packageVersion("RColorBrewer")
library("dunn.test"); packageVersion("dunn.test")
library("DESeq2"); packageVersion("DESeq2")
setwd("./data")
opts_knit$set(root.dir = "./data")
#opts_chunk$set(tidy.opts=list(width.cutoff=80))
```

### Import data

First the matrix percent file and count file generated by the minimum entropy decomposition (MED) pipeline, subsampled to 7974 reads per sample, and the accociated taxonony file  


```{r, tidy=TRUE, tidy.opts=list(width.cutoff=80)}
allShared = read.table("all.7974.matrixPercent.txt", header=T, row.names=1)
allCounts = read.table("all.7974.matrixCount.txt", header=T, row.names=1)
allTax = read.table('all.7974.nodeReps.nr_v119.knn.taxonomy', header=T, sep='\t', row.names = 1)
allTax = allTax[,2:8]
allTax = as.matrix(allTax)
```

Import the shared and taxonomy files generated in mothur for 3% and 1% pairwise similarity, in order to calculate alpha diversity measures and to compare to the MED procedure. Also import the 3% OTU file without any subsampling for alpha diversity calculations.   


```{r}
all3OTUshared = read.table("all.7974.0.03.pick.shared", header=T, row.names=2)
all3OTUshared = all3OTUshared[,3:length(all3OTUshared)]

alpha3OTUshared = read.table("all.7974.0.03.shared", header=T)
rownames(alpha3OTUshared) = alpha3OTUshared[,2]
alpha3OTUshared = alpha3OTUshared[,4:length(alpha3OTUshared)]

all1OTUshared = read.table("all.7974.0.01.pick.shared", header=T, row.names=2)
all1OTUshared = all1OTUshared[,3:length(all1OTUshared)]

all3OTUtax = read.table('all.7974.0.03.taxonomy', header=T, sep='\t', row.names=1)
all3OTUtax = all3OTUtax[,2:8]
all3OTUtax = as.matrix(all3OTUtax)

all1OTUtax = read.table('all.7974.0.01.taxonomy', header=T, sep='\t', row.names=1)
all1OTUtax = all1OTUtax[,2:8]
all1OTUtax = as.matrix(all1OTUtax)
```

Import Endozoicomonas phylogenetic tree (exported from ARB) using the APE package (Fig. 3). Also import a MED percent matrix that is slightly modified to accomodate the tree  


```{r}
endoTreeFile = read.tree(file='MEDNJ5.tree')
allSharedTree = read.table("all.7974.matrixPercent.tree.txt", header=T, row.names=1)
```


Import meta data for the samples, including metaData3.txt, which is slightly modified to accomodate heatmap sample ordering, and metaDataChem which contains additional columns of physiochemical data  

```{r}
metaFile = read.table('metaData2.MED', header=T, sep='\t', row.names=1)
metaFile3 = read.table('metaData3.txt', header=T, sep='\t', row.names=1)
metaFileChem = read.table('metaDataChem.txt', header=T, sep='\t', row.names=1)
```

### Create phyloseq objects and add consistent coloring for sites

```{r, tidy=TRUE, tidy.opts=list(width.cutoff=80)}
OTU = otu_table(allShared, taxa_are_rows = FALSE)
OTUcounts = otu_table(allCounts, taxa_are_rows = FALSE)
OTUs3 = otu_table(all3OTUshared, taxa_are_rows = FALSE)
OTUs3alpha = otu_table(alpha3OTUshared, taxa_are_rows = FALSE)
OTUs1 = otu_table(all1OTUshared, taxa_are_rows = FALSE)
OTUtree = otu_table(allSharedTree, taxa_are_rows = FALSE)

TAX = tax_table(allTax) 
TAX3 = tax_table(all3OTUtax)
TAX1 = tax_table(all1OTUtax)

META = sample_data(metaFile)
METAchem = sample_data(metaFileChem)
TREE = phy_tree(endoTreeFile)

allPhylo = phyloseq(OTU, TAX, META)
countPhylo = phyloseq(OTUcounts, TAX, META)
all3OTUphylo = phyloseq(OTUs3, TAX3, META)
alpha3OTUphylo = phyloseq(OTUs3alpha, META)
all1OTUphylo = phyloseq(OTUs1, TAX1, META)
allPhyloChem = phyloseq(OTU, TAX, METAchem)
endoTree = phyloseq(OTUtree, META, TREE)

cols <- c("AmericanSamoa" = "#D95F02", "Indonesia" = "#A6761D", "MaggieIs" = "#666666", "Maldives" = "#E6AB02", "Micronesia" = "#66A61E", "Ningaloo" = "#7570B3", "RedSea" = "#E7298A", "other" = "black")
```

### Ordinations to compare MED vs pairwise OTUs 

Subset samples for the two corals, remove taxa with 0s, create relative abundance and square-root sample counts

```{r}
filter_stylo_data <- function(initial_matrix){
  initial_coral <- subset_samples(initial_matrix, species=="Stylophora pistillata")
  coral_filt = filter_taxa(initial_coral, function(x) mean(x) > 0, TRUE)
  coral_filt_rel = transform_sample_counts(coral_filt, function(x) x / sum(x) )
  coral_filt_rel_sqrt = transform_sample_counts(coral_filt_rel, function(x) sqrt(x) ) 
  return(coral_filt_rel_sqrt)
}

filter_pverr_data <- function(initial_matrix){
  initial_coral <- subset_samples(initial_matrix, species=="Pocillopora verrucosa")
  coral_filt = filter_taxa(initial_coral, function(x) mean(x) > 0, TRUE)
  coral_filt_rel = transform_sample_counts(coral_filt, function(x) x / sum(x) )
  coral_filt_rel_sqrt = transform_sample_counts(coral_filt_rel, function(x) sqrt(x) ) 
  return(coral_filt_rel_sqrt)
}

spistPhyloRelSqrt <- filter_stylo_data(allPhylo)
spist3OTUphyloRelSqrt <- filter_stylo_data(all3OTUphylo)
spist1OTUphyloRelSqrt <- filter_stylo_data(all1OTUphylo)

pverrPhyloRelSqrt <- filter_pverr_data(allPhylo)
pverr3OTUphyloRelSqrt <- filter_pverr_data(all3OTUphylo)
pverr1OTUphyloRelSqrt <- filter_pverr_data(all1OTUphylo)
```

Now do ordinations for each 

```{r, tidy=TRUE, tidy.opts=list(width.cutoff=80)}
compOrdinations <- function(sample_data, sample_name){
theme_set(theme_bw())
sample_dataOrd <- ordinate(sample_data, "NMDS", "bray")
plot_ordination(sample_data, sample_dataOrd, type = 'samples', color='site', title=sample_name) +
  geom_point(size=2) +
  scale_color_manual(values=cols)
}

compOrdinations(spistPhyloRelSqrt, "S. pistillata MED OTUs")
compOrdinations(spist3OTUphyloRelSqrt, "S. pistillata 3% OTUs")
compOrdinations(spist1OTUphyloRelSqrt, "S. pistillata 1% OTUs")

compOrdinations(pverrPhyloRelSqrt, "P. verrucosa MED OTUs")
compOrdinations(pverr3OTUphyloRelSqrt, "P. verrucosa 3% OTUs")
compOrdinations(pverr1OTUphyloRelSqrt, "P. verrucosa 1% OTUs")
```

### Alpha diversity measures

First subset the corals, then plot using phyloseq and ggplot2

Note: I'll use unsubampled 3% pairwise OTUs for calculation of alpha diversity measures as this will make them more comparable to other studies, plus the MED pipeline is has not yet implemented alpha diversity

```{r, tidy=TRUE, tidy.opts=list(width.cutoff=80)}
allAlphaTmp <- subset_samples(alpha3OTUphylo, species=="seawater")
allAlphaTmp2 <- subset_samples(alpha3OTUphylo, species=="Stylophora pistillata")
allAlphaTmp3 <- subset_samples(alpha3OTUphylo, species=="Pocillopora verrucosa")
allAlpha2 <- merge_phyloseq(allAlphaTmp, allAlphaTmp2, allAlphaTmp3)

allAlphaPlot2 <- plot_richness(allAlpha2, x = 'species', measures = c('Chao1', 'Simpson', 'observed'), color = 'site', sortby = 'Chao1')

ggplot(data = allAlphaPlot2$data) +
  geom_point(aes(x = species, y = value, color = site), position=position_jitter(width=0.1, height=0)) +
  geom_boxplot(aes(x = species, y = value, color = NULL), alpha = 0.1, outlier.shape = NA) +
  scale_color_manual(values=cols) +
  theme(axis.text.x = element_text(angle = 90)) +
  facet_wrap(~variable, scales='free_y') +
  scale_x_discrete(limits=c("Stylophora pistillata", "Pocillopora verrucosa", "seawater"))
```

Check for significant differences in the alpha diversity measures using a kruskal-wallis test and a dunn post-hoc test to check which specific groups were different

```{r}
alphaObserved = estimate_richness(allAlpha2, measures="Observed")
alphaSimpson = estimate_richness(allAlpha2, measures="Simpson")
alphaChao = estimate_richness(allAlpha2, measures="Chao1")

alpha.stats <- cbind(alphaObserved, sample_data(allAlpha2))
alpha.stats2 <- cbind(alpha.stats, alphaSimpson)
alpha.stats3 <- cbind(alpha.stats2, alphaChao)

kruskal.test(Observed~species, data = alpha.stats3)
dunn.test(alpha.stats3$Observed, alpha.stats3$species, method="bonferroni")

kruskal.test(Simpson~species, data = alpha.stats3)
dunn.test(alpha.stats3$Simpson, alpha.stats3$species, method="bonferroni")

kruskal.test(Chao1~species, data = alpha.stats3)
dunn.test(alpha.stats3$Chao1, alpha.stats3$species, method="bonferroni")
```

In each case, the seawater was signficiantly different to the corals, while the corals were not different to each other. This suggests the corals have a more 'selective' community of microbes compared to the surrounding seawater. 

### Similarity Profile Analysis (SIMPROF)

This will show how the samples cluster without any a priori assumptions regarding sample origin

Need to import the shared file containing just spist OTUs, then calcualte the simprof clusters based on the braycurtis metric. 

```{r, fig.width = 10, tidy=TRUE, tidy.opts=list(width.cutoff=80)}
spist <- subset_samples(allPhylo, species=='Stylophora pistillata')
spistShared = otu_table(spist)
class(spistShared) <- "numeric"

spistSIMPROF <- simprof(spistShared, num.expected=1000, num.simulated=99, method.cluster='average', method.distance='braycurtis', method.transform='squareroot', alpha=0.05, sample.orientation='row', silent=TRUE)

simprof.plot(spistSIMPROF, leafcolors=NA, plot=TRUE, fill=TRUE, leaflab="perpendicular", siglinetype=1)

pVerr <- subset_samples(allPhylo, species=='Pocillopora verrucosa')
pVerrShared = otu_table(pVerr)
class(pVerrShared) <- "numeric"

pVerrSIMPROF <- simprof(pVerrShared, num.expected=1000, num.simulated=99, method.cluster='average', method.distance='braycurtis', method.transform='squareroot', alpha=0.05, sample.orientation='row', silent=TRUE)

simprof.plot(pVerrSIMPROF, leafcolors=NA, plot=TRUE, fill=TRUE, leaflab="perpendicular", siglinetype=1)
```

### Chemical and biological correlations

Use the envfit function from the Vegan package to test if any environmental variables are significantly correlated with microbiome differences in the corals

```{r, tidy=TRUE, tidy.opts=list(width.cutoff=80)}
draw_envfit_ord <- function(coral_chem, env_data){   
  chemNoNA <- na.omit(metaFileChem[sample_names(coral_chem),env_data])
  coralNoNA <- prune_samples(rownames(chemNoNA), coral_chem)
  
  theme_set(theme_bw())
coralNoNAOrd <- ordinate(coralNoNA, "NMDS", "bray")
coralNoNAOrdPlot <- plot_ordination(coralNoNA, coralNoNAOrd, type = 'samples', color='site') +
  geom_point(size=3) +
  scale_color_manual(values=c(cols)) 
  
  # get points for ggplot
  pointsNoNA <- coralNoNAOrd$points[rownames(chemNoNA),]
  chemFit <- envfit(pointsNoNA, env = chemNoNA, na.rm=TRUE)
  print(chemFit)
  chemFit.scores <- as.data.frame(scores(chemFit, display= "vectors"))
  chemFit.scores <- cbind(chemFit.scores, Species = rownames(chemFit.scores))

  # create arrow info
chemNames <- rownames(chemFit.scores)  
arrowmap <- aes(xend = MDS1, yend = MDS2, x = 0, y = 0, shape = NULL, color = NULL)
  labelmap <- aes(x = MDS1, y = MDS2 + 0.04, shape = NULL, color = NULL, size=1.5, label = chemNames)
  arrowhead = arrow(length = unit(0.25, "cm"))
  
  # note: had to use aes_string to get ggplot to recognize variables
coralNoNAOrdPlot + 
  coord_fixed() +
  geom_segment(arrowmap, size = 0.5, data = chemFit.scores, color = "black",  arrow = arrowhead, show_guide = FALSE) +
  geom_text(aes_string(x = "MDS1", y = "MDS2", shape = NULL, color = NULL, size=1.5, label = "Species"), size = 3, data = chemFit.scores)
}

waterQual <- c("temp", "salinity", "Domg", "pH")
nutrients <- c("PO4", "N.N", "silicate", "NO2", "NH4")
FCM <- c("prok", "syn", "peuk", "pe.peuk", "Hbact")

spistChem <- subset_samples(allPhyloChem, species=='Stylophora pistillata')
pverrChem <- subset_samples(allPhyloChem, species=='Pocillopora verrucosa')

draw_envfit_ord(spistChem, waterQual)
draw_envfit_ord(spistChem, nutrients)
draw_envfit_ord(spistChem, FCM)

draw_envfit_ord(pverrChem, waterQual)
draw_envfit_ord(pverrChem, nutrients)
draw_envfit_ord(pverrChem, FCM)
```

### Taxonomic barcharts of bacteria in the corals and seawaters, and core microbiome members

```{r, fig.width = 10, fig.height=6, tidy=TRUE, tidy.opts=list(width.cutoff=80)}
# define a function to draw barcharts at a specific taxonomic level
# also need to create my own ggplot colors then replace the last one ('other' column) with gray

gg_color_hue <- function(n) {
  hues = seq(15, 375, length=n+1)
  hcl(h=hues, l=65, c=100)[1:n]
}

draw_barcharts <- function(coral_species, tax_level) {

coralFiltGlom <- tax_glom(coral_species, taxrank=tax_level)
physeqdf <- psmelt(coralFiltGlom)

# get total abundance so can make an 'other' column
# had to add ^ and $ characters to make sure grep matches whole word

physeqdfOther <- physeqdf

for (j in unique(physeqdf$Sample)) {
  jFirst = paste('^', j, sep='')
  jBoth = paste(jFirst, '$', sep='')
  rowNumbers = grep(jBoth, physeqdf$Sample)
  otherValue = 100 - sum(physeqdf[rowNumbers,"Abundance"])
  newRow = (physeqdf[rowNumbers,])[1,]
  newRow[,tax_level] = "other"
  newRow[,"Abundance"] = otherValue
  physeqdfOther <- rbind(physeqdfOther, newRow)
}

ggCols <- gg_color_hue(length(unique(physeqdfOther[,tax_level])))
ggCols <- head(ggCols, n=-1)

# add names and numbers for easier referencing
physeqdfOther$names <- factor(physeqdfOther$Sample, levels=rownames(metaFile), ordered = TRUE)
physeqdfOther$tax_level_num <- as.numeric(physeqdfOther[,tax_level])

theme_set(theme_bw())
ggplot(physeqdfOther, aes_string(x="names", y="Abundance", fill=tax_level, order=tax_level)) +
  geom_bar(stat="identity", colour="black") +
  #geom_text(position = 'stack', aes(label = tax_level_num, vjust = 2, y = Abundance), size = 3) +
  geom_text(position = 'stack', aes(label = ifelse(Abundance>2, tax_level_num,''), vjust = 1.5, y = check_abund(Abundance)), size = 3) +
  scale_fill_manual(values=c(ggCols, "gray")) +
  scale_y_continuous(expand = c(0,0), limits = c(0,100)) +
  facet_grid(~site, scales='free', space='free_x') +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
}

# subset coral samples, create names factor for label ordering and filter so the graph isn't too full
spist <- subset_samples(allPhylo, species=='Stylophora pistillata')
sample_data(spist)$names <- factor(sample_names(spist), levels=rownames(metaFile), ordered = TRUE)
spistFilt = filter_taxa(spist, function(x) mean(x) > 0.8, TRUE)
draw_barcharts(spistFilt, "Phylum") # 0.2
draw_barcharts(spistFilt, "Class") # 0.5
draw_barcharts(spistFilt, "Genus") # 0.8

pVerr <- subset_samples(allPhylo, species=='Pocillopora verrucosa')
sample_data(pVerr)$names <- factor(sample_names(pVerr), levels=unique(sample_names(pVerr)))
pVerrFilt = filter_taxa(pVerr, function(x) mean(x) > 0.3, TRUE)
draw_barcharts(pVerrFilt, "Phylum")
draw_barcharts(pVerrFilt, "Class")
draw_barcharts(pVerrFilt, "Genus")

sea <- subset_samples(allPhylo, species=='seawater')
sample_data(sea)$names <- factor(sample_names(sea), levels=rownames(metaFile), ordered = TRUE)
seaFilt = filter_taxa(sea, function(x) mean(x) > 0.3, TRUE)
draw_barcharts(seaFilt, "Phylum")
draw_barcharts(seaFilt, "Class")
draw_barcharts(seaFilt, "Family")
```

The two corals are both dominated by Gammaproteobacteria at the higher taxonomic levels. At the genus level, there is more variability but Endozoicomonas seem to be fairly prevalent. Let's check which bacterial genera are most consistently associated with the corals and may be considered a 'core' microbiome member.

### Core coral microbiome members

```{r, tidy=TRUE, tidy.opts=list(width.cutoff=80)}
# check for 'core' microbiome members at the genus level
# which taxa are present at 1% overall abundance and at least 50% of samples in Stylophora pistillata?
spistGenusGlom <- tax_glom(spistFilt, taxrank="Genus")
coreTaxa = filter_taxa(spistGenusGlom, function(x) sum(x > 1) > (0.5*length(x)), TRUE)
tax_table(coreTaxa)
sum(otu_table(coreTaxa) > 1) / nsamples(spist)

# which taxa are present at 1% overall abundance and at least 50% of samples in Pocillopora verrucosa?
pVerrGenusGlom <- tax_glom(pVerrFilt, taxrank="Genus")
coreTaxa = filter_taxa(pVerrGenusGlom, function(x) sum(x > 1) > (0.5*length(x)), TRUE)
tax_table(coreTaxa)
sum(otu_table(coreTaxa) > 1) / nsamples(pVerr)
```

Indeed Endozoicomonas were the most prevalent bacteria in the corals and were they only bacterial genera to occur in more than 50% of the colonies sampled. In fact, for both coral species Endozoicomonas occurred in more than 75% of colonies.

Let's check if these Endozoicomonas OTUs show any patterns across the coral species or at different geographic areas.

### Heatmap of different Endozoicomonas MED OTUs across the coral species

```{r, fig.width = 10, fig.height=8, tidy=TRUE, tidy.opts=list(width.cutoff=80)}
allPhyloEndo <- subset_taxa(allPhylo, Genus=='Endozoicomonas')
spistPhyloEndo <- subset_samples(allPhyloEndo, species=='Stylophora pistillata')
pVerrPhyloEndo <- subset_samples(allPhyloEndo, species=='Pocillopora verrucosa')
spistPverrEndo <- merge_phyloseq(spistPhyloEndo, pVerrPhyloEndo)

spistPverrEndoFilt = filter_taxa(spistPverrEndo, function(x) mean(x) > 0.0, TRUE)
spistPverrEndoFiltPrune = prune_samples(sample_sums(spistPverrEndoFilt) > 0, spistPverrEndoFilt)

plot_heatmap(spistPverrEndoFiltPrune, "NMDS", "bray", "site", low="#000033", high="#FF3300", sample.order=rownames(metaFile3))
plot_heatmap(spistPverrEndoFiltPrune, "NMDS", "bray", "species", low="#000033", high="#FF3300", sample.order=rownames(metaFile3))
```

Pretty cool. Looks like the two corals have different Endozoicomonas types, and the types also seem to partition differently across sites for the coral species, i.e., Pocillopora verrucosa has similar Endozoicomonas types across large geographic areas but Stylophora pistillata seems to have different Endozoicomonas types at each area. Let's do some significance testing to see if more of the Endozoicomonas OTUs are different across sites for S. pistillata compared to P. verrucosa.

### DESeq2 significance testing for Endozoicomonas MED OTUs

```{r}
# subset endozoicomonas OTUs from the count data as required by DESeq2
countEndos <- subset_taxa(countPhylo, Genus=='Endozoicomonas')
spistCountEndo <- subset_samples(countEndos, species=='Stylophora pistillata')
pVerrCountEndo <- subset_samples(countEndos, species=='Pocillopora verrucosa')
coralCountEndo <- merge_phyloseq(spistCountEndo, pVerrCountEndo)

# do some filtering for 0s
spistCountEndoFilt = filter_taxa(spistCountEndo, function(x) mean(x) > 0.0, TRUE)
spistCountEndoFiltPrune = prune_samples(sample_sums(spistCountEndoFilt) > 0, spistCountEndoFilt)
pVerrCountEndoFilt = filter_taxa(pVerrCountEndo, function(x) mean(x) > 0.0, TRUE)
pVerrCountEndoFiltPrune = prune_samples(sample_sums(pVerrCountEndoFilt) > 0, pVerrCountEndoFilt)

# convert phyloseq object to DESeq object
spistDeseq <- phyloseq_to_deseq2(spistCountEndoFiltPrune, ~ site)
pverrDeseq <- phyloseq_to_deseq2(pVerrCountEndoFiltPrune, ~ site)

# need to calculate geometric means separately because there are zeros in the data
gm_mean = function(x, na.rm=TRUE){
  exp(sum(log(x[x > 0]), na.rm=na.rm) / length(x))
}
spistMeans <- apply(counts(spistDeseq), 1, gm_mean)
spistDeseq <- estimateSizeFactors(spistDeseq, geoMeans = spistMeans)
pverrMeans <- apply(counts(pverrDeseq), 1, gm_mean)
pverrDeseq <- estimateSizeFactors(pverrDeseq, geoMeans = pverrMeans)
 
# now can run the DESeq tests
spistDeseq <- DESeq(spistDeseq, fitType="local")
pverrDeseq <- DESeq(pverrDeseq, fitType="local")

# create a function to check the results for each of the site comparisons
# output significant p-value (FDR adjusted) ordered OTUs

check_site_results <- function(coral_deseq, site_comparison){
  res <- results(coral_deseq, contrast = site_comparison)
  res = res[order(res$padj, na.last=NA), ]
  sigtab = as.data.frame(res[(res$padj < 0.05), ])
  print(sigtab)
}

# Stylophora pistillata
check_site_results(spistDeseq, c("site", "AmericanSamoa", "RedSea"))
check_site_results(spistDeseq, c("site", "AmericanSamoa", "Ningaloo"))
check_site_results(spistDeseq, c("site", "AmericanSamoa", "Micronesia"))
check_site_results(spistDeseq, c("site", "AmericanSamoa", "Indonesia"))
check_site_results(spistDeseq, c("site", "RedSea", "Ningaloo"))
check_site_results(spistDeseq, c("site", "RedSea", "Micronesia"))
check_site_results(spistDeseq, c("site", "RedSea", "Indonesia"))
check_site_results(spistDeseq, c("site", "Ningaloo", "Micronesia"))
check_site_results(spistDeseq, c("site", "Ningaloo", "Indonesia"))
check_site_results(spistDeseq, c("site", "Micronesia", "Indonesia"))

# Pocillopora verrucosa
check_site_results(pverrDeseq, c("site", "Maldives", "RedSea"))
check_site_results(pverrDeseq, c("site", "Maldives", "Micronesia"))
check_site_results(pverrDeseq, c("site", "Maldives", "Indonesia"))
check_site_results(pverrDeseq, c("site", "RedSea", "Micronesia"))
check_site_results(pverrDeseq, c("site", "RedSea", "Indonesia"))
check_site_results(pverrDeseq, c("site", "Micronesia", "Indonesia"))
```

There is quite a few more significant differences across sites for Stylophora pistillata (90) compared to Pocillopora verrucosa (18), supporting what is shown in the heatmaps. I'll add an asterisk next to each of the significantly different OTUs in the heatmap to visually display these results. 

Endozoicomonas seem to be displaying strain-specific relationships with the corals and across sites. I'll do a phylogenetic analysis of the Endozoicomonas sequences to further explore these relationships. The sequences were aligned using the SINA web service and imported into ARB for manual refinement, before being exported for use here. 

### Endozoicomonas phylogenetic tree with meta-data

```{r, fig.width=10, fig.height=10, tidy=TRUE, tidy.opts=list(width.cutoff=80)}
# subset out our corals / seawater of interest
endoTreeSpist <- subset_samples(endoTree, species=='Stylophora pistillata')
endoTreePverr <- subset_samples(endoTree, species=='Pocillopora verrucosa')
endoTreeSea <- subset_samples(endoTree, species=='seawater')
endoTreeOther <- subset_samples(endoTree, species=='other')
endoTreeCorals <- merge_phyloseq(endoTreeSpist, endoTreePverr, endoTreeSea, endoTreeOther)

# plot tree - phyloseq makes this easy

plot_tree(endoTreeCorals, label.tips = 'taxa_names', color='site', shape='species', size='abundance', nodelabf = nodeplotboot(100, 50, 3), ladderize='left', base.spacing = 0.01) +
  scale_color_manual(values=cols) +
  scale_shape_manual(values = c("other" = 1, "Pocillopora verrucosa" = 17, "Stylophora pistillata" = 16, "seawater" = 15)) 
```

Some interesting things coming up here. Several abundant but phylogenetically diverse Endozoicomonas OTUs appear to co-inhabit the same coral colonies. There are also clear host and site groupings of Endozoicomonas strains. I also included single cell 16S sequences in the tree and they are identical to the abundant MED OTUs from the Red Sea, suggesting that the MED procedure produces biologically relevant OTUs. Also in the tree are sequences from an earlier study of Red Sea Stylophora pistillata, named 'Spistillata_17_6_E02, Spistillata_12_6_A02, Spistillata_15_6_D04', and these are also the same as the abundant MED nodes in this study, suggesting that the most abundant MED OTUs in Red Sea S. pistillata have not changed for ~ 4 years.






