---
title: "coralMicrobiomes"
author: "Matthew J. Neave"
date: "April 22, 2015"
output: html_document
---


This is an R Markdown document detailing the statistical and graphical steps for reproducting the results in:

Neave, M.J., Rachmawati, R., Xun, L., Michell, C.T., Barber, P.H., Bourne, D.G., McCulloch, M.T., Apprill, A., Voolstra, C.R. A global microbiome analysis reveals that closely related corals exhibit fine-scale differences in their association with Endozoicomonas symbionts. 

### Load required libraries  


```r
library("phyloseq"); packageVersion("phyloseq")
```

```
## [1] '1.9.15'
```

```r
library("ggplot2"); packageVersion("ggplot2")
```

```
## [1] '1.0.0'
```

```r
library("plyr"); packageVersion("plyr")
```

```
## [1] '1.8.1'
```

```r
library("vegan"); packageVersion("vegan")
```

```
## [1] '2.0.10'
```

```r
library("grid"); packageVersion("grid")
```

```
## [1] '3.1.1'
```

```r
library("knitr"); packageVersion("knitr")
```

```
## [1] '1.6'
```

```r
library("clustsig"); packageVersion("clustsig")
```

```
## Warning: package 'clustsig' was built under R version 3.1.2
```

```
## [1] '1.1'
```

```r
library('ape'); packageVersion("ape")
```

```
## [1] '3.1.4'
```

```r
library('RColorBrewer'); packageVersion("RColorBrewer")
```

```
## [1] '1.0.5'
```

```r
library("dunn.test"); packageVersion("dunn.test")
```

```
## Warning: package 'dunn.test' was built under R version 3.1.2
```

```
## [1] '1.2.3'
```

```r
library("DESeq2"); packageVersion("DESeq2")
```

```
## Warning: package 'RcppArmadillo' was built under R version 3.1.2
```

```
## [1] '1.4.5'
```

```r
setwd("./data")
opts_knit$set(root.dir = "./data")
```

### Import data

First the matrix percent file generated by the minimum entropy decomposition (MED) pipeline, subsampled to 7974 reads per sample, and the accociated taxonony file  



```r
allShared = read.table("all.7974.matrixPercent.txt", header=T, row.names=1)
allTax = read.table('all.7974.nodeReps.nr_v119.knn.taxonomy', header=T, sep='\t', row.names = 1)
```

```
## Warning: number of items read is not a multiple of the number of columns
```

```r
allTax = allTax[,2:8]
allTax = as.matrix(allTax)
```

Import the shared and taxonomy files generated in mothur for 3% and 1% pairwise similarity, in order to calculate alpha diversity measures and to compare to the MED procedure. Also import the 3% OTU file without any subsampling for alpha diversity calculations.   



```r
all3OTUshared = read.table("all.7974.0.03.pick.shared", header=T, row.names=2)
all3OTUshared = all3OTUshared[,3:length(all3OTUshared)]

alpha3OTUshared = read.table("all.7974.0.03.shared", header=T)
rownames(alpha3OTUshared) = alpha3OTUshared[,2]
alpha3OTUshared = alpha3OTUshared[,4:length(alpha3OTUshared)]

all1OTUshared = read.table("all.7974.0.01.pick.shared", header=T, row.names=2)
all1OTUshared = all1OTUshared[,3:length(all1OTUshared)]

all3OTUtax = read.table('all.7974.0.03.taxonomy', header=T, sep='\t', row.names=1)
all3OTUtax = all3OTUtax[,2:8]
all3OTUtax = as.matrix(all3OTUtax)

all1OTUtax = read.table('all.7974.0.01.taxonomy', header=T, sep='\t', row.names=1)
all1OTUtax = all1OTUtax[,2:8]
all1OTUtax = as.matrix(all1OTUtax)
```

Import Endozoicomonas phylogenetic tree (exported from ARB) using the APE package (Fig. 3). Also import a MED percent matrix that is slightly modified to accomodate the tree  



```r
endoTreeFile = read.tree(file='MEDNJ5.tree')
allSharedTree = read.table("all.7974.matrixPercent.tree.txt", header=T, row.names=1)
```


Import meta data for the samples, including metaData3.txt, which is slightly modified to accomodate heatmap sample ordering, and metaDataChem which contains additional columns of physiochemical data  


```r
metaFile = read.table('metaData2.MED', header=T, sep='\t', row.names=1)
metaFile3 = read.table('metaData3.txt', header=T, sep='\t', row.names=1)
metaFileChem = read.table('metaDataChem.txt', header=T, sep='\t', row.names=1)
```

### Create phyloseq objects and add consistent coloring for sites


```r
OTU = otu_table(allShared, taxa_are_rows = FALSE)
OTUs3 = otu_table(all3OTUshared, taxa_are_rows = FALSE)
OTUs3alpha = otu_table(alpha3OTUshared, taxa_are_rows = FALSE)
OTUs1 = otu_table(all1OTUshared, taxa_are_rows = FALSE)
OTUtree = otu_table(allSharedTree, taxa_are_rows = FALSE)

TAX = tax_table(allTax) 
TAX3 = tax_table(all3OTUtax)
TAX1 = tax_table(all1OTUtax)

META = sample_data(metaFile)
METAchem = sample_data(metaFileChem)
TREE = phy_tree(endoTreeFile)

allPhylo = phyloseq(OTU, TAX, META)
all3OTUphylo = phyloseq(OTUs3, TAX3, META)
alpha3OTUphylo = phyloseq(OTUs3alpha, META)
all1OTUphylo = phyloseq(OTUs1, TAX1, META)
allPhyloChem = phyloseq(OTU, TAX, METAchem)
endoTree = phyloseq(OTUtree, META, TREE)

cols <- c("AmericanSamoa" = "#D95F02", "Indonesia" = "#A6761D", "MaggieIs" = "#666666", "Maldives" = "#E6AB02", "Micronesia" = "#66A61E", "Ningaloo" = "#7570B3", "RedSea" = "#E7298A")
```

### Ordinations to compare MED vs pairwise OTUs 

Subset samples for the two corals, remove taxa with 0s, create relative abundance and square-root sample counts


```r
filter_stylo_data <- function(initial_matrix){
  initial_coral <- subset_samples(initial_matrix, species=="Stylophora pistillata")
  coral_filt = filter_taxa(initial_coral, function(x) mean(x) > 0, TRUE)
  coral_filt_rel = transform_sample_counts(coral_filt, function(x) x / sum(x) )
  coral_filt_rel_sqrt = transform_sample_counts(coral_filt_rel, function(x) sqrt(x) ) 
  return(coral_filt_rel_sqrt)
}

filter_pverr_data <- function(initial_matrix){
  initial_coral <- subset_samples(initial_matrix, species=="Pocillopora verrucosa")
  coral_filt = filter_taxa(initial_coral, function(x) mean(x) > 0, TRUE)
  coral_filt_rel = transform_sample_counts(coral_filt, function(x) x / sum(x) )
  coral_filt_rel_sqrt = transform_sample_counts(coral_filt_rel, function(x) sqrt(x) ) 
  return(coral_filt_rel_sqrt)
}

spistPhyloRelSqrt <- filter_stylo_data(allPhylo)
spist3OTUphyloRelSqrt <- filter_stylo_data(all3OTUphylo)
spist1OTUphyloRelSqrt <- filter_stylo_data(all1OTUphylo)

pverrPhyloRelSqrt <- filter_pverr_data(allPhylo)
pverr3OTUphyloRelSqrt <- filter_pverr_data(all3OTUphylo)
pverr1OTUphyloRelSqrt <- filter_pverr_data(all1OTUphylo)
```

Now do ordinations for each 


```r
compOrdinations <- function(sample_data, sample_name){
theme_set(theme_bw())
sample_dataOrd <- ordinate(sample_data, "NMDS", "bray")
plot_ordination(sample_data, sample_dataOrd, type = 'samples', color='site', title=sample_name) +
  geom_point(size=2) +
  scale_color_manual(values=cols)
}

compOrdinations(spistPhyloRelSqrt, "S. pistillata MED OTUs")
```

```
## Run 0 stress 0.2254 
## Run 1 stress 0.2459 
## Run 2 stress 0.2413 
## Run 3 stress 0.2399 
## Run 4 stress 0.2389 
## Run 5 stress 0.2373 
## Run 6 stress 0.2344 
## Run 7 stress 0.2328 
## Run 8 stress 0.2371 
## Run 9 stress 0.229 
## Run 10 stress 0.2273 
## Run 11 stress 0.2341 
## Run 12 stress 0.2302 
## Run 13 stress 0.2352 
## Run 14 stress 0.2288 
## Run 15 stress 0.2503 
## Run 16 stress 0.2381 
## Run 17 stress 0.2486 
## Run 18 stress 0.2409 
## Run 19 stress 0.2242 
## ... New best solution
## ... procrustes: rmse 0.02453  max resid 0.1974 
## Run 20 stress 0.2292
```

<img src="coralMicrobiomes_files/figure-html/unnamed-chunk-71.png" title="plot of chunk unnamed-chunk-7" alt="plot of chunk unnamed-chunk-7" width="672" />

```r
compOrdinations(spist3OTUphyloRelSqrt, "S. pistillata 3% OTUs")
```

```
## Run 0 stress 0.2273 
## Run 1 stress 0.235 
## Run 2 stress 0.248 
## Run 3 stress 0.2386 
## Run 4 stress 0.2395 
## Run 5 stress 0.2663 
## Run 6 stress 0.2542 
## Run 7 stress 0.2273 
## ... New best solution
## ... procrustes: rmse 0.0004645  max resid 0.003106 
## *** Solution reached
```

<img src="coralMicrobiomes_files/figure-html/unnamed-chunk-72.png" title="plot of chunk unnamed-chunk-7" alt="plot of chunk unnamed-chunk-7" width="672" />

```r
compOrdinations(spist1OTUphyloRelSqrt, "S. pistillata 1% OTUs")
```

```
## Run 0 stress 0.2225 
## Run 1 stress 0.2575 
## Run 2 stress 0.2512 
## Run 3 stress 0.2274 
## Run 4 stress 0.2645 
## Run 5 stress 0.2512 
## Run 6 stress 0.2366 
## Run 7 stress 0.2274 
## Run 8 stress 0.2351 
## Run 9 stress 0.2532 
## Run 10 stress 0.2304 
## Run 11 stress 0.2582 
## Run 12 stress 0.2338 
## Run 13 stress 0.2318 
## Run 14 stress 0.2331 
## Run 15 stress 0.2399 
## Run 16 stress 0.2586 
## Run 17 stress 0.2571 
## Run 18 stress 0.2417 
## Run 19 stress 0.2287 
## Run 20 stress 0.2559
```

<img src="coralMicrobiomes_files/figure-html/unnamed-chunk-73.png" title="plot of chunk unnamed-chunk-7" alt="plot of chunk unnamed-chunk-7" width="672" />

```r
compOrdinations(pverrPhyloRelSqrt, "P. verrucosa MED OTUs")
```

```
## Run 0 stress 0.2437 
## Run 1 stress 0.2485 
## Run 2 stress 0.271 
## Run 3 stress 0.2362 
## ... New best solution
## ... procrustes: rmse 0.09863  max resid 0.3381 
## Run 4 stress 0.2446 
## Run 5 stress 0.2349 
## ... New best solution
## ... procrustes: rmse 0.07606  max resid 0.353 
## Run 6 stress 0.2133 
## ... New best solution
## ... procrustes: rmse 0.08037  max resid 0.3474 
## Run 7 stress 0.2259 
## Run 8 stress 0.2227 
## Run 9 stress 0.2438 
## Run 10 stress 0.2136 
## ... procrustes: rmse 0.02494  max resid 0.146 
## Run 11 stress 0.2247 
## Run 12 stress 0.2454 
## Run 13 stress 0.2186 
## Run 14 stress 0.2393 
## Run 15 stress 0.2393 
## Run 16 stress 0.2206 
## Run 17 stress 0.2315 
## Run 18 stress 0.2264 
## Run 19 stress 0.2572 
## Run 20 stress 0.239
```

<img src="coralMicrobiomes_files/figure-html/unnamed-chunk-74.png" title="plot of chunk unnamed-chunk-7" alt="plot of chunk unnamed-chunk-7" width="672" />

```r
compOrdinations(pverr3OTUphyloRelSqrt, "P. verrucosa 3% OTUs")
```

```
## Run 0 stress 0.2409 
## Run 1 stress 0.2354 
## ... New best solution
## ... procrustes: rmse 0.09003  max resid 0.3035 
## Run 2 stress 0.2482 
## Run 3 stress 0.2598 
## Run 4 stress 0.2349 
## ... New best solution
## ... procrustes: rmse 0.09954  max resid 0.4044 
## Run 5 stress 0.23 
## ... New best solution
## ... procrustes: rmse 0.1085  max resid 0.3136 
## Run 6 stress 0.2283 
## ... New best solution
## ... procrustes: rmse 0.09872  max resid 0.2776 
## Run 7 stress 0.224 
## ... New best solution
## ... procrustes: rmse 0.0909  max resid 0.2662 
## Run 8 stress 0.2289 
## Run 9 stress 0.2407 
## Run 10 stress 0.2288 
## Run 11 stress 0.2238 
## ... New best solution
## ... procrustes: rmse 0.04183  max resid 0.2614 
## Run 12 stress 0.2593 
## Run 13 stress 0.2852 
## Run 14 stress 0.2277 
## Run 15 stress 0.227 
## Run 16 stress 0.2358 
## Run 17 stress 0.2468 
## Run 18 stress 0.2353 
## Run 19 stress 0.2316 
## Run 20 stress 0.4043
```

<img src="coralMicrobiomes_files/figure-html/unnamed-chunk-75.png" title="plot of chunk unnamed-chunk-7" alt="plot of chunk unnamed-chunk-7" width="672" />

```r
compOrdinations(pverr1OTUphyloRelSqrt, "P. verrucosa 1% OTUs")
```

```
## Run 0 stress 0.234 
## Run 1 stress 0.2289 
## ... New best solution
## ... procrustes: rmse 0.08099  max resid 0.2296 
## Run 2 stress 0.2458 
## Run 3 stress 0.217 
## ... New best solution
## ... procrustes: rmse 0.09259  max resid 0.4139 
## Run 4 stress 0.2297 
## Run 5 stress 0.2492 
## Run 6 stress 0.2473 
## Run 7 stress 0.228 
## Run 8 stress 0.2341 
## Run 9 stress 0.2263 
## Run 10 stress 0.2306 
## Run 11 stress 0.2187 
## Run 12 stress 0.2263 
## Run 13 stress 0.2331 
## Run 14 stress 0.2356 
## Run 15 stress 0.2355 
## Run 16 stress 0.2172 
## ... procrustes: rmse 0.04413  max resid 0.297 
## Run 17 stress 0.2188 
## Run 18 stress 0.2283 
## Run 19 stress 0.2181 
## Run 20 stress 0.2322
```

<img src="coralMicrobiomes_files/figure-html/unnamed-chunk-76.png" title="plot of chunk unnamed-chunk-7" alt="plot of chunk unnamed-chunk-7" width="672" />

### Alpha diversity measures

First subset the corals, then plot using phyloseq and ggplot2

Note: I'll use unsubampled 3% pairwise OTUs for calculation of alpha diversity measures as this will make them more comparable to other studies, plus the MED pipeline is has not yet implemented alpha diversity


```r
allAlphaTmp <- subset_samples(alpha3OTUphylo, species=="seawater")
allAlphaTmp2 <- subset_samples(alpha3OTUphylo, species=="Stylophora pistillata")
allAlphaTmp3 <- subset_samples(alpha3OTUphylo, species=="Pocillopora verrucosa")
allAlpha2 <- merge_phyloseq(allAlphaTmp, allAlphaTmp2, allAlphaTmp3)

allAlphaPlot2 <- plot_richness(allAlpha2, x = 'species', measures = c('Chao1', 'Simpson', 'observed'), color = 'site', sortby = 'Chao1')

ggplot(data = allAlphaPlot2$data) +
  geom_point(aes(x = species, y = value, color = site), position=position_jitter(width=0.1, height=0)) +
  geom_boxplot(aes(x = species, y = value, color = NULL), alpha = 0.1, outlier.shape = NA) +
  scale_color_manual(values=cols) +
  theme(axis.text.x = element_text(angle = 90)) +
  facet_wrap(~variable, scales='free_y') +
  scale_x_discrete(limits=c("Stylophora pistillata", "Pocillopora verrucosa", "seawater"))
```

```
## Warning: Removed 3 rows containing missing values (geom_point).
## Warning: Removed 7 rows containing missing values (geom_point).
## Warning: Removed 4 rows containing missing values (geom_point).
## Warning: Removed 6 rows containing missing values (geom_point).
## Warning: Removed 4 rows containing missing values (geom_point).
## Warning: Removed 7 rows containing missing values (geom_point).
## Warning: Removed 4 rows containing missing values (geom_point).
## Warning: Removed 4 rows containing missing values (geom_point).
```

<img src="coralMicrobiomes_files/figure-html/unnamed-chunk-8.png" title="plot of chunk unnamed-chunk-8" alt="plot of chunk unnamed-chunk-8" width="672" />

Check for significant differences in the alpha diversity measures using a kruskal-wallis test and a dunn post-hoc test to check which specific groups were different


```r
alphaObserved = estimate_richness(allAlpha2, measures="Observed")
alphaSimpson = estimate_richness(allAlpha2, measures="Simpson")
alphaChao = estimate_richness(allAlpha2, measures="Chao1")

alpha.stats <- cbind(alphaObserved, sample_data(allAlpha2))
alpha.stats2 <- cbind(alpha.stats, alphaSimpson)
alpha.stats3 <- cbind(alpha.stats2, alphaChao)

kruskal.test(Observed~species, data = alpha.stats3)
```

```
## 
## 	Kruskal-Wallis rank sum test
## 
## data:  Observed by species
## Kruskal-Wallis chi-squared = 61.88, df = 2, p-value = 3.662e-14
```

```r
dunn.test(alpha.stats3$Observed, alpha.stats3$species, method="bonferroni")
```

```
##   Kruskal-Wallis rank sum test
## 
## data: x and group
## Kruskal-Wallis chi-squared = 61.8764, df = 2, p-value = 0
## 
## 
##                            Comparison of x by group                            
##                                  (Bonferroni)                                  
## Col Mean-|
## Row Mean |   Pocillop   seawater
## ---------+----------------------
## seawater |   7.510384
##          |     0.0000
##          |
## Stylopho |   1.357184  -6.783011
##          |     0.2621     0.0000
```

```r
kruskal.test(Simpson~species, data = alpha.stats3)
```

```
## 
## 	Kruskal-Wallis rank sum test
## 
## data:  Simpson by species
## Kruskal-Wallis chi-squared = 12.25, df = 2, p-value = 0.002193
```

```r
dunn.test(alpha.stats3$Simpson, alpha.stats3$species, method="bonferroni")
```

```
##   Kruskal-Wallis rank sum test
## 
## data: x and group
## Kruskal-Wallis chi-squared = 12.2453, df = 2, p-value = 0
## 
## 
##                            Comparison of x by group                            
##                                  (Bonferroni)                                  
## Col Mean-|
## Row Mean |   Pocillop   seawater
## ---------+----------------------
## seawater |   3.397898
##          |     0.0010
##          |
## Stylopho |   0.811204  -2.904738
##          |     0.6259     0.0055
```

```r
kruskal.test(Chao1~species, data = alpha.stats3)
```

```
## 
## 	Kruskal-Wallis rank sum test
## 
## data:  Chao1 by species
## Kruskal-Wallis chi-squared = 64.31, df = 2, p-value = 1.086e-14
```

```r
dunn.test(alpha.stats3$Chao1, alpha.stats3$species, method="bonferroni")
```

```
##   Kruskal-Wallis rank sum test
## 
## data: x and group
## Kruskal-Wallis chi-squared = 64.3067, df = 2, p-value = 0
## 
## 
##                            Comparison of x by group                            
##                                  (Bonferroni)                                  
## Col Mean-|
## Row Mean |   Pocillop   seawater
## ---------+----------------------
## seawater |   7.581725
##          |     0.0000
##          |
## Stylopho |   1.146749  -7.033279
##          |     0.3772     0.0000
```

In each case, the seawater was signficiantly different to the corals, while the corals were not different to each other. This suggests the corals have a more 'selective' community of microbes compared to the surrounding seawater. 

### Similarity Profile Analysis (SIMPROF)

This will show how the samples cluster without any a priori assumptions regarding sample origin

Need to import the shared file containing just spist OTUs, then calcualte the simprof clusters based on the braycurtis metric. 


```r
# spist <- subset_samples(allPhylo, species=='Stylophora pistillata')
# spistShared = otu_table(spist)
# class(spistShared) <- "numeric"
# 
# spistSIMPROF <- simprof(spistShared, num.expected=1000, num.simulated=999, method.cluster='average', method.distance='braycurtis', method.transform='squareroot', alpha=0.05, sample.orientation='row', silent=FALSE)
# 
# simprof.plot(spistSIMPROF, leafcolors=NA, plot=TRUE, fill=TRUE, leaflab="perpendicular", siglinetype=1)
# 
# pVerr <- subset_samples(allPhylo, species=='Pocillopora verrucosa')
# pVerrShared = otu_table(pVerr)
# class(pVerrShared) <- "numeric"
# 
# pVerrSIMPROF <- simprof(pVerrShared, num.expected=1000, num.simulated=999, method.cluster='average', method.distance='braycurtis', method.transform='squareroot', alpha=0.05, sample.orientation='row', silent=FALSE)
# 
# simprof.plot(pVerrSIMPROF, leafcolors=NA, plot=TRUE, fill=TRUE, leaflab="perpendicular", siglinetype=1)
```





